runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Timer.java

```diff
@@ -40,92 +52,201 @@
   }
 
+  /** Returns the key that the timer is set on. */
+  public abstract T getUserKey();
+
+  /**
+   * Returns the tag that the timer is set on. This tag is {@code ""} when the timer is for a {@link
+   * TimerSpec}.
+   */
+  public abstract String getDynamicTimerTag();
+
+  /** Returns the windows which is associated with the timer. */
+  public abstract Collection<? extends BoundedWindow> getWindows();
+
+  /** Returns whether the timer is going to be cleared. */
+  public abstract boolean getClearBit();
+
   /**
-   * Returns the timestamp of when the timer is scheduled to fire.
+   * Returns the timestamp of when the timer is scheduled to fire. This field is nullable only when
+   * the timer is being cleared.
    *
-   * <p>The time is relative to the time domain defined in the {@link
+   * <p>The time is absolute to the time domain defined in the {@link
    * org.apache.beam.model.pipeline.v1.RunnerApi.TimerFamilySpec} that is associated with this
    * timer.
    */
-  public abstract Instant getTimestamp();
+  @Nullable
+  public abstract Instant getFireTimestamp();
 
-  /** A user supplied payload. */
+  /**
+   * Returns the watermark that the timer is supposed to be held. This field is nullable only when
+   * the timer is being cleared.
+   */
   @Nullable
-  public abstract T getPayload();
+  public abstract Instant getHoldTimestamp();
+
+  /**
+   * Returns the paneinfo that is related to the timer. This field is nullable only when the
+   * timer is being cleared.
+   */
+  @Nullable
+  public abstract PaneInfo getPane();
+
+  @Override
+  public boolean equals(Object other) {
+    if (!(other instanceof Timer)) {
+      return false;
+    }
+    Timer<?> that = (Timer<?>) other;
+    if (this.getClearBit()) {
+      return Objects.equals(this.getUserKey(), that.getUserKey())
+          && this.getDynamicTimerTag().equals(that.getDynamicTimerTag())
+          && (this.getClearBit() == that.getClearBit())
+          && Objects.equals(this.getWindows(), that.getWindows());
+    }
+    return Objects.equals(this.getUserKey(), that.getUserKey())
+        && this.getDynamicTimerTag().equals(that.getDynamicTimerTag())
+        && (this.getClearBit() == that.getClearBit())
+        && this.getFireTimestamp().equals(that.getFireTimestamp())
+        && this.getHoldTimestamp().equals(that.getHoldTimestamp())
+        && Objects.equals(this.getWindows(), that.getWindows())
+        && Objects.equals(this.getPane(), that.getPane());
+  }
+
+  @Override
+  public int hashCode() {
+    // Hash only the millis of the timestamp to be consistent with equals
+    if (getClearBit()) {
+      return Objects.hash(getUserKey(), getDynamicTimerTag(), getClearBit(), getWindows());
+    }
+    return Objects.hash(
+        getUserKey(),
+        getDynamicTimerTag(),
+        getClearBit(),
+        getFireTimestamp().getMillis(),
+        getHoldTimestamp().getMillis(),
+        getWindows(),
+        getPane());
```

Comment:
```suggestion
    return Objects.hash(
        getUserKey(),
        getDynamicTimerTag(),
        getWindows(),
        getClearBit(),
        getFireTimestamp().getMillis(),
        getHoldTimestamp().getMillis(),
        getPane());
```

Reply:
`getFireTimestamp()` returns `null` when `clearBit` is `true`. I think we still need to check  `clearBit` in order to hash the `timestamp` with milliseconds.

Reply:
good point